neural_speed/models/stablelm/stablelm.cpp

```diff
@@ -0,0 +1,428 @@
+          ne_build_forward_expand(&gf, ne_flash_attn_update_k(ctx0, k_cache, Kcur, n_past, false));
+          const auto v_cache = ne_view_3d(ctx0, kv_self.v,          // tensor
+                                          head_dim, n_ctx, n_head,  // ne
+                                          0, 0,                     // nb (bestla managed)
+                                          il * v_size);             // offset
+          ne_build_forward_expand(&gf, ne_flash_attn_update_v(ctx0, v_cache, Vcur, n_past, false));
+        }
+
+        struct ne_tensor* Q = ne_permute(ctx0, Qcur, 0, 2, 1, 3);
+        ne_set_name(Q, "Q");
+
+        struct ne_tensor* K =
+            ne_view_3d(ctx0, kv_self.k,                                             // tensor
+                       head_dim, seq_kv, n_head,                                    // ne
+                       kv_cache_info.stride_k_sl, kv_cache_info.stride_k_head_num,  // nb (bestla managed)
+                       il * k_size);                                                // offset
+        *reinterpret_cast<ATTN_FWD_LAYOUT*>(&K->nb[0]) = kv_cache_info.k_layout;    // us nb0 for layout
+        ne_set_name(K, "K");
+        struct ne_tensor* V =
+            ne_view_3d(ctx0, kv_self.v,                                                    // tensor
+                       seq_kv, head_dim, n_head,                                           // ne
+                       kv_cache_info.stride_v_head_size, kv_cache_info.stride_v_head_num,  // nb (bestla managed)
+                       il * v_size);                                                       // offset
+        *reinterpret_cast<ATTN_FWD_LAYOUT*>(&V->nb[0]) = kv_cache_info.v_layout;           // us nb0 for layout
+        ne_set_name(V, "V");
+
+        ne_attn_flags_t attn_flags = 0;
+        if (n_past == 0) attn_flags |= NE_ATTN_FLAG_IS_CAUSAL;  // no causal mask on next-token cases
+        struct ne_tensor* KQV_Out = ne_flash_attn(ctx0, Q, K, V, attn_scale, attn_flags);
+        cur = ne_view_2d(ctx0, KQV_Out, n_embd, N, n_embd * ne_element_size(KQV_Out), 0);
+      }
+
+      // projection
+      {
+        if (n_layer == 24) {  // Stablelm2 1.6B & Stablelm2 Zephyr 1.6B
+          cur = ne_mul_mat(ctx0, model.layers[il].attn[6], cur);
+        } else {  // Stablelm 3B
+          cur = ne_mul_mat(ctx0, model.layers[il].attn[3], cur);
+        }
+      }
+    }
+    lctx.use_buf(ctx0, 1);
+
+    cur = ne_add(ctx0, cur, inpL);
+    inpL = cur;
+
+    // FFN Block
+    {
+      // Post Attention norm
+      {
+        cur = ne_norm(ctx0, cur, hparams.norm_eps);
+        cur = ne_mul(ctx0, cur, model.layers[il].norm[2]);
+        cur = ne_add(ctx0, cur, model.layers[il].norm[3]);
+      }
+
+      if (bestla_fusion_FFN_SiLu_f32f32_support(model.layers[il].ffn[0]->data, model.layers[il].ffn[1]->data,
+                                                model.layers[il].ffn[2]->data, N, cur->ne[0],
```

Comment:
have you test on SPR with MHA FFN fusion is enabled?

Reply:
In your doc, you say MHA fusion is supported only for AMX_BF16. I tested with --compute_dtype int8, so to enable the MHA fusion I assume I need to switch the computation to BF16 ?

Reply:
As long as you are inferencing on SPR, the mha fusion is default enabled.

Reply:
@a32543254 Then the answer to your question is yes. The results above I shared are on SPR dual socket. Btw, I was wondering is there a way to have report like you did on Phi models, I'm speaking about this: https://inteltf-jenk.sh.intel.com/job/neural_speed_extension/47/artifact/report.html